yt/frontends/nc4_cm1/io.py

```diff
@@ -0,0 +1,62 @@
+import numpy as np
+
+from yt.utilities.file_handler import NetCDF4FileHandler
+from yt.utilities.io_handler import BaseIOHandler
+
+
+class CM1IOHandler(BaseIOHandler):
+    _particle_reader = False
+    _dataset_type = "cm1"
+
+    def __init__(self, ds):
+        self.filename = ds.filename
+        self._handle = NetCDF4FileHandler(self.filename)
+        super(CM1IOHandler, self).__init__(ds)
+
+    def _read_particle_coords(self, chunks, ptf):
+        # This needs to *yield* a series of tuples of (ptype, (x, y, z)).
+        # chunks is a list of chunks, and ptf is a dict where the keys are
+        # ptypes and the values are lists of fields.
+        pass
+
+    def _read_particle_fields(self, chunks, ptf, selector):
+        # This gets called after the arrays have been allocated.  It needs to
+        # yield ((ptype, field), data) where data is the masked results of
+        # reading ptype, field and applying the selector to the data read in.
+        # Selector objects have a .select_points(x,y,z) that returns a mask, so
+        # you need to do your masking here.
+        pass
```

Comment:
To other reviewers: should these functions be:

- like they currently are
- raise a NotImplementedError
- be removed

Also, this goes well beyond the scope of this PR, but is `_read_particle_coords` still a thing? It seems to be only used for SPH-specific datasets.

Reply:
For now I advise `raise NotImplementedError`. If CM1 dataset never have particles however they can be removed.

> Also, this goes well beyond the scope of this PR, but is _read_particle_coords still a thing? It seems to be only used for SPH-specific datasets.

Yeah that's one of the reasons why we should have a cleaner class separation for dataset types.

Reply:
We do have particles, but they are contained in separate NetCDF files. I'd lean towards the NotImplementedError for that reason, as we may try to add that in a future pull request